subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy

```diff
@@ -59,9 +61,219 @@ class SwiftIncrementalCompileIntegrationTest extends AbstractInstalledToolChainI
+        buildFile << """
+            tasks.withType(SwiftCompile) {
+                compilerArgs.add('-Onone')
+            }
+        """
+
+        expect:
+        succeeds("compileDebugSwift")
+        outputs.recompiledClasses('main', 'sum', 'greeter', 'multiply')
+    }
+
+    def 'changes to an unused dependency rebuilds everything'() {
+        given:
+        def outputs = new CompilationOutputsFixture(file("build/obj/main/debug"), [ ".o" ])
+        def app = new SwiftApp()
+        settingsFile << """
+            rootProject.name = 'app'
+            include 'unused'
+        """
+        buildFile << """
+            apply plugin: 'swift-application'
+            
+            project(":unused") {
+                apply plugin: 'swift-library'
+            }
+        """
+        file("unused/src/main/swift/Library.swift") << """
+            public class Library {
+                public init() {}
+                public func unused() {
+                }
+            }
+        """
+        app.writeToProject(testDirectory)
+
+        outputs.snapshot { succeeds("compileDebugSwift") }
+
+        buildFile << """
+            application {
+                dependencies {
+                    if (project.hasProperty("includeDep")) {
+                        implementation(project(":unused"))
+                    }
+                }
+            }
+        """
+
+        expect:
+        // Addition of 'unused' dependency rebuilds everything.
+        succeeds("compileDebugSwift", "-PincludeDep")
+        outputs.recompiledClasses('main', 'sum', 'greeter', 'multiply')
+
+        and:
+        // Removal of 'unused' dependency rebuilds everything.
+        succeeds("compileDebugSwift")
+        outputs.recompiledClasses('main', 'sum', 'greeter', 'multiply')
+    }
+
+    @RequiresInstalledToolChain(ToolChainRequirement.SWIFTC_4)
+    def 'changing Swift language level rebuilds everything'() {
+        given:
+        def outputs = new CompilationOutputsFixture(file("build/obj/main/debug"), [ ".o" ])
+        def app = new SwiftApp()
+        settingsFile << "rootProject.name = 'app'"
+        app.writeToProject(testDirectory)
+        buildFile << """
+            apply plugin: 'swift-application'
+            application {
+                if (project.hasProperty("swift4")) {
+                    sourceCompatibility = SwiftVersion.SWIFT4
+                } else {
+                    sourceCompatibility = SwiftVersion.SWIFT3
+                }
+            }
+         """
+
+        outputs.snapshot { succeeds("compileDebugSwift") }
+
+        expect:
+        // rebuild for Swift4
+        succeeds("compileDebugSwift", "-Pswift4")
+        outputs.recompiledClasses('main', 'sum', 'greeter', 'multiply')
+
+        and:
+        // rebuild for Swift3
+        succeeds("compileDebugSwift")
+        outputs.recompiledClasses('main', 'sum', 'greeter', 'multiply')
+    }
+
+    @RequiresInstalledToolChain(ToolChainRequirement.SWIFTC_4)
```

Comment:
This doesn't seem quite correct.  It's not that we require swift4 - we require both swift3 and swift4.  This probably works for all of our installation scenarios, but it's ripe for flakiness.  What we ultimately need is a new annotation that looks at all toolchains and checks against multiple `ToolChainRequirement`s (or modify `RequiresInstalledToolChain` to accept multiple requirements.  I'd be happy with a comment here for now that this is an approximation and what we might do to make it better, or an issue, or both.

Reply:
Yeah, I only added it so it wouldn't run twice. I'll add a comment